### PR TITLE
Proxy Resolver

### PR DESCRIPTION
**Pull request by tjwiebell**

### CODE REVIEW

File: src/Resolver/Proxy.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        return 'proxy';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isValid(Definition $definition): bool
+    {
+        $rootDefinition = $this->getIterator()->getRootDefinition();
```
Comment (bbatsche) [inline src/Resolver/Proxy.php]: Why are you looking at the root definition and validating a key from there? Should you be looking at the definition passed and checking it's required properties?

File: src/Resolver/Proxy.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function resolve($definition)
+    {
+        $target          = $this->getIterator()->get('proxy.target');
```
Comment (bbatsche) [inline src/Resolver/Proxy.php]: Again, should be working off of the passed `$definition` not looking at some other key in the root.

File: src/Resolver/Proxy.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     * {@inheritdoc}
+     */
+    public function isValid(Definition $definition): bool
+    {
+        $rootDefinition = $this->getIterator()->getRootDefinition();
+        if ($rootDefinition->has('proxy')) {
+            $proxyDefinition = $rootDefinition->get('proxy');
+            if (!$proxyDefinition->has('target')) {
+                return false;
+            }
```
Comment (bbatsche) [inline src/Resolver/Proxy.php]: I'd add some quick validation for `ignoreSSLErrors` to make sure it's boolean

File: src/Resolver/Proxy.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+use Zend\Http\Client;
+
+class Proxy extends AbstractResolver
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getIndicator(): string
+    {
+        return 'proxy';
```
Comment (bbatsche) [inline src/Resolver/Proxy.php]: Indicator should be `target` not `proxy`: https://github.com/magento-research/pwa-studio/tree/release/2.0/packages/upward-spec#resolver-type-inference

File: test/Resolver/ProxyTest.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        $this->definitionIteratorMock->shouldReceive('get')
+            ->with('proxy.target')
+            ->andReturn('https://google.com');
+        $this->definitionIteratorMock->shouldReceive('getRootDefinition')->andReturn($rootDefinition);
+        $this->definitionIteratorMock->shouldReceive('get')
+            ->with('proxy.ignoreSSLErrors')
+            ->andReturn(true);
+
+        $zendClientMock = Mockery::mock('overload:' . Client::class);
+        $zendClientMock->shouldReceive('send');
```
Comment (bbatsche) [inline test/Resolver/ProxyTest.php]: `$zendClientMock` should return a `Response` object, and we should verify that `resolve()` is returning that same response.

File: src/Resolver/Proxy.php (hunk 1 above)
Comment (tjwiebell) [reply src/Resolver/Proxy.php]: My mistake, I think you're correct. In looking at the Venia configs, I mistook `use: proxy` in conditional resolvers as a static root level config all proxies shared. Paired with my misunderstanding of the indicator too, this makes a lot more sense now, good catch.

File: src/Template/TemplateFactory.php [hunk 6]
```diff
@@ -20,6 +20,8 @@ class TemplateFactory
     ];
 
     /**
+     * @param ?string $engine
```
Comment (bbatsche) [inline src/Template/TemplateFactory.php]: I don't know why PHP CS Fixer keeps insisting on adding this!

Comment (bbatsche) [inline src/Template/TemplateFactory.php]: I don't know why PHP CS Fixer keeps insisting on adding this!

Comment (bbatsche) [reply src/Template/TemplateFactory.php]: I mean, _do_ know, but it's wrong so it's frustrating.

**Pull request merged**